lgc/builder/DescBuilder.cpp

```diff
@@ -109,29 +109,71 @@ Value *DescBuilder::CreateLoadBufferDesc(unsigned descSet, unsigned binding, Val
   }
 
   if (!desc) {
-    // Not handled by either of the special cases above...
-    // Get a pointer to the descriptor, as a pointer to i8.
-    // For shader compilation with no user data layout provided, we assume we want a DescriptorBuffer, as
-    // DescriptorConstBuffer is not used in that case.
-    ResourceNodeType resType = node ? node->concreteType : ResourceNodeType::DescriptorBuffer;
-    ResourceNodeType abstractType = node ? node->abstractType : resType;
-    Value *descPtr = getDescPtr(resType, abstractType, descSet, binding, topNode, node);
-    // Index it.
-    if (descIndex != getInt32(0)) {
-      descIndex = CreateMul(descIndex, getStride(resType, descSet, binding, node));
-      descPtr = CreateGEP(getInt8Ty(), descPtr, descIndex);
+    if (node) {
+      ResourceNodeType resType = node->concreteType;
+      ResourceNodeType abstractType = node->abstractType;
+      Value *descPtr = getDescPtr(resType, abstractType, descSet, binding, topNode, node);
+      // Index it.
+      if (descIndex != getInt32(0)) {
+        descIndex = CreateMul(descIndex, getStride(resType, descSet, binding, node));
+        descPtr = CreateGEP(getInt8Ty(), descPtr, descIndex);
+      }
+      // Cast it to the right type.
+      descPtr = CreateBitCast(descPtr, getDescPtrTy(resType));
+      // Load the descriptor.
+      desc = CreateLoad(getDescTy(resType), descPtr);
+    } else {
+      // If descriptor set is -1, this is a internal resource node, it is a root node
+      // and its type is ResourceNodeType::DescriptorBufferCompact.
+      if (descSet == -1)
+        return loadInternalBuffer(descSet, binding, pointeeTy, instName);
```

Comment:
Why does this receive special treatment? It honestly just seems like bad code duplication.

If we know for a fact that these buffers are always compact, then the better way to deal with it is inline, here, by skipping the pieces of code that are not required.